File: tictactoe.cpp
#include <iostream>

const char empty = '-';
char ttt[3][3];

bool checkHorizontal() {
    if(((ttt[0][0]==ttt[0][1])&&(ttt[0][1]==ttt[0][2])&&(ttt[0][1]!=empty))||((ttt[1][0]==ttt[1][1])&&(ttt[1][1]==ttt[1][2])&&(ttt[1][1]!=empty))||((ttt[2][0]==ttt[2][1])&&(ttt[2][1]==ttt[2][2])&&(ttt[2][2]!=empty)))
        return true;
    else
        return false;
}

bool checkVertical() {
    if(((ttt[0][0]==ttt[1][0])&&(ttt[1][0]==ttt[2][0])&&(ttt[0][0]!=empty))||((ttt[0][1]==ttt[1][1])&&(ttt[1][1]==ttt[2][1])&&(ttt[0][1]!=empty))||((ttt[0][2]==ttt[1][2])&&(ttt[1][2]==ttt[2][2])&&(ttt[0][2]!=empty)))
        return true;
    else
        return false;
}

bool checkDiagonal() {
    if(((ttt[0][0]==ttt[1][1])&&(ttt[1][1]==ttt[2][2])&&(ttt[0][0]!=empty))||((ttt[0][2]==ttt[1][1])&&(ttt[1][1]==ttt[2][0])&&(ttt[1][1]!=empty)))
        return true;
    else
        return false;
}

bool checkWin() {
    if(checkHorizontal() || checkVertical() || checkDiagonal())
        return true;
    else
        return false;
}

void showVector() {
    system("clear");
    for(int t=0;t<3;t++) {
        std::cout<<ttt[t][0]<<" | "<<ttt[t][1]<<" | "<<ttt[t][2]<<std::endl;
    }
}

void user() {
    int userx, usery;
    std::cout << "Enter the coordinate to put 'X'";
    std::cin >> userx >> usery;
    if ((userx < 0 || userx > 2) || (usery < 0 || usery > 2)) {
        std::cout << "Please enter a valid input!!" << std::endl;
        user();
    }
    else if(ttt[userx][usery] != empty) {
        std::cout << "Cell is not empty!" << std::endl;
        user();
    }
    else {
        ttt[userx][usery] = 'X';
        if(checkWin()) {
            showVector();
            std::cout<<"You won!"<<std::endl;
            exit(EXIT_SUCCESS);
        } else
        showVector();
    }
}

void comp() {
    int compx, compy;
    compx = rand() % 3;
    compy = rand() % 3;
    if(ttt[compx][compy] != empty)
        comp();
    else {
        ttt[compx][compy] = 'O';
        if(checkWin()) {
            showVector();
            std::cout<<"Computer won!"<<std::endl;
            exit(EXIT_SUCCESS);
        } else
        showVector();
    }
}

int main() {
    int count=0;
    for(int i=0;i<3;i++) {
        for(int j=0;j<3;j++) {
            ttt[i][j] = empty;
        }
    }
    showVector();
    while(true) {
        if(count%2==0) {
            std::cout<<"Your Turn"<<std::endl;
            user();
            count++;
        }
        else {
            std::cout<<"Computer Turn"<<std::endl;
            comp();
            count++;
        }
    }
    return 0;
}

File: tictactoe.go
package main

import ("fmt"
	"math/rand"
	"os"
	"os/exec"
)

const empty string = "-"
var ttt[3][3] string

func checkHorizontal() bool {
	if ((ttt[0][0] == ttt[0][1]) && (ttt[0][1] == ttt[0][2]) && (ttt[0][1] != empty)) || ((ttt[1][0] == ttt[1][1]) && (ttt[1][1] == ttt[1][2]) && (ttt[1][1] != empty)) || ((ttt[2][0] == ttt[2][1]) && (ttt[2][1] == ttt[2][2]) && (ttt[2][2] != empty)) {
		return true
	}
	return false
}

func checkVertical() bool {
	if ((ttt[0][0] == ttt[1][0]) && (ttt[1][0] == ttt[2][0]) && (ttt[0][0] != empty)) || ((ttt[0][1] == ttt[1][1]) && (ttt[1][1] == ttt[2][1]) && (ttt[0][1] != empty)) || ((ttt[0][2] == ttt[1][2]) && (ttt[1][2] == ttt[2][2]) && (ttt[0][2] != empty)) {
		return true
	}
	return false
}

func checkDiagonal() bool {
	if ((ttt[0][0] == ttt[1][1]) && (ttt[1][1] == ttt[2][2]) && (ttt[0][0] != empty)) || ((ttt[0][2] == ttt[1][1]) && (ttt[1][1] == ttt[2][0]) && (ttt[1][1] != empty)) {
		return true
	}
	return false
}

func checkWin() bool {
	if checkHorizontal() || checkVertical() || checkDiagonal() {
		return true
	}
	return false
}

func showVector() {
	c := exec.Command("clear")
	c.Stdout = os.Stdout
	c.Run()
	for t:=0; t<3; t++ {
	fmt.Println(ttt[t][0], " | ", ttt[t][1], " | ", ttt[t][2])
	}
}

func user() {
	var userx, usery int
	fmt.Println("Enter the coordinate to put 'X'")
	fmt.Scan(&userx, &usery)
	if (userx < 0 || userx > 2) || (usery < 0 || usery > 2) {
		fmt.Println("Please enter a valid input!!")
		user()
	} else if ttt[userx][usery] != empty {
		fmt.Println("Cell is not empty!")
		user()
	} else {
		ttt[userx][usery] = "X"
		if checkWin() {
			showVector()
			fmt.Println("You won!")
			os.Exit(0)
		} else {
			showVector()
		}
	}
}

func comp() {
	var compx= rand.Intn(3)
	var compy= rand.Intn(3)
	if ttt[compx][compy] != empty {
		comp()
	} else {
		ttt[compx][compy] = "O"
		if checkWin() {
			showVector()
			fmt.Println("Computer won!")
			os.Exit(0)
		} else {
			showVector()
		}
	}
}

func main() {

	count := 0
	for i:=0; i<3; i++ {
		for j:=0; j<3; j++ {
			ttt[i][j] = empty
		}
	}
	showVector()
	for true {
		if count%2==0 {
			fmt.Println("Your Turn")
			user()
			count++
		} else {
			fmt.Println("Computer's Turn")
			comp()
			count++
		}
	}
}
